PR: Remove trailing spaces and add script/module to automate the process

CAM-Gerlach opened the pull request:
<!--- Make sure to read the Contributing Guidelines:                   --->
<!--- https://github.com/spyder-ide/spyder/blob/master/CONTRIBUTING.md --->
<!--- and follow PEP 8, PEP 257 and Spyder's code style:               --->
<!--- https://github.com/spyder-ide/spyder/wiki/Dev:-Coding-Style      --->

## Description of Changes

* [x] Wrote at least one-line docstrings (for any new functions)

<!--- Explain what you've done and why --->

In two separate commits, as requested, this PR removes trailing whitespace on both blank, and non-blank lines, from files with extensions on a manually curated whitelist, and with exclusions for any non-code test files and the vendored mathjax library. Also, the changes are checked against every open PR from the past >6 months, and any files that have changes that result in merge conflicts are excluded. They comprise the following:

<details>

```
PR #9226 has conflicts {'spyder/plugins/onlinehelp/onlinehelp.py', 'spyder/plugins/profiler/plugin.py', 'spyder/plugins/workingdirectory/plugin.py', 'spyder/plugins/findinfiles/plugin.py', 'spyder/plugins/breakpoints/plugin.py', 'spyder/plugins/console/plugin.py', 'spyder/plugins/pylint/plugin.py', 'spyder/plugins/outlineexplorer/plugin.py', 'spyder/plugins/variableexplorer/plugin.py', 'spyder/plugins/history/plugin.py', 'spyder/plugins/explorer/plugin.py'}
PR #9219 has conflicts {'spyder/plugins/outlineexplorer/widgets.py'}
PR #9127 has conflicts {'spyder/utils/workers/types.py'}
PR #9082 has conflicts {'spyder/plugins/outlineexplorer/widgets.py', 'spyder/utils/syntaxhighlighters.py'}
PR #9014 has conflicts {'spyder/plugins/outlineexplorer/widgets.py', 'spyder/utils/syntaxhighlighters.py'}
PR #8852 has conflicts {'spyder/plugins/variableexplorer/widgets/namespacebrowser.py'}
Total PR conflicts: 15 in 6
```

</details>

Finally, this work adds a comprehensive [script and module](https://github.com/spyder-ide/spyder/pull/9307/files#diff-71288fc955fd5bb4ed90b1943621795b), which can be accessed via APIs for both the command line and from other Python programs, that does the checks. It can be configured to either actually execute the changes, or merely warn about them and fail a build accordingly; and it can also optionally either scan the entire index, or various configurable subsets of modified files for use as a manual ``make check`` style script or an automated pre-commit hook. Many common components and infrastructure can likely be re-used in the other checker/fixer scripts, and it has been quite considerably tested over various scenarios (though by no means exhaustively).

Ideally, if #9226  could get merged quickly, that would take care of over 2/3rds of the remaining merge conflicts and I could re-base this one accordingly. Then, only 5 files remain, which could be handled on a PR by PR basis. #8734 would also be useful to get merged soon to help keep things clean until the final CI checks, pre-commit hooks and/or cleanup script can get implemented.


### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

<!--- TYPE YOUR USER/NAME AFTER THE FOLLOWING: --->
I certify the above statement is true and correct: CAM-Gerlach

<!--- Thanks for your help making Spyder better for everyone! --->

ccordoba12 commented:
@CAM-Gerlach, thanks a lot for your work on this!! I'm leaving this beta3 because we're trying to have everything ready to release beta2 next week.

We'll come back to review this afterwards (there are some whitespace removals that are unnecessary and the script should't be placed in `scripts`).

CAM-Gerlach commented:
> I'm leaving this beta3 because we're trying to have everything ready to release beta2 next week.

Okay sure. If conflicts develop, I can always just regenerate it with the script.

> there are some whitespace removals that are unnecessary

Aren't *all* of them technically unnecessary)? I thought the point was to be consistent with removing them everywhere, so we can keep the "remove trailing whitespace on save" feature on and run ``yapf``, this script, etc to keep things clean without worrying about causing future conflicts.

> the script should't be placed in scripts

That's just where @goanpeca suggested I put it. Where should it go instead?

ccordoba12 commented:
> Aren't all of them technically unnecessary?

No, I already noted a few that are necessary, so I need to review this PR with care.

ccordoba12 commented:
> That's just where @goanpeca suggested I put it. Where should it go instead?

Ok, but the scripts in `scripts` are used to launch Spyder on Linux.

So it perhaps should go to `tools` in the root of the repo. And you can also move there `png-convert-and-optimize.sh`, which has been in the wrong place for a long time, but I didn't know where to put it.

CAM-Gerlach commented:
> No, I already noted a few that are necessary

> there are some whitespace removals that are unnecessary

... :confounded: 

> Ok, but the scripts in `scripts` are used to launch Spyder on Linux.

Yeah, it did seem a little odd to put it there. There are also some random Windows scripts too.

> And you can also move there `png-convert-and-optimize.sh

Yes, that makes sense to put it wherever this goes. It definitely doesn't belong in the repo root like you said. We can put the other scripts and stuff in there so we can have a script to call everything at once. We could even integrate ``png_convert_and_optimize`` into the that architecture and automatically call it for any new or changed PNGs, along with our other stuff. Then link it all to a git pre-commit hook and never have to worry about it again...

ccordoba12 commented:
> Then link it all to a git pre-commit hook and never have to worry about it again...

Yeah, that'd be nice too.

It's your choice then for the new directory name. My suggestion are `tools` or `tooling`. Scikit-learn has `maint_tools`, which is good too.

CAM-Gerlach commented:
> Yeah, that'd be nice too.

Out of scope here, ofc :) But that's the next step after we have the scripts working and everything harmonized.

> It's your choice then for the new directory name

I probably don't have any better suggestions than you, lol. @goanpeca , any thoughts?

goanpeca commented:
Yes my bad, in my head and other repos the `scripts` folder is used for these. So hmmmmm I would go with `tools` for now and maybe at a `README.md` as well ? 

>So it perhaps should go to tools in the root of the repo. And you can also move there png-convert-and-optimize.sh, which has been in the wrong place for a long time, but I didn't know where to put it.

👍 

Thanks for working on this :-)

goanpeca commented:
I will get back to review in more detail after beta2 is released! This is looking great 🎉

goanpeca commented:
@CAM-Gerlach could you update this?

@ccordoba12 could you point to some of the spaces that were not supposed to be removed to check what the issue might be?

CAM-Gerlach commented:
The only one I immediately see are the deliberate spaces in the issue template. Other than that, I'm not sure exactly what he meant, as I was rather conservative with this (not applying it to any non-actual-test files in the test directory, for instance, or any file types not on an explicit whitelist). However, its easy to add any files/paths that should not be converted to the blacklist in the script.

I can re-run it, but I'd rather wait to know what exclusions to add, since with the various Github calls, test-merging every PR and the substantial increase in unmerged PRs lately (with a number about to be merged), the runtime and resource cost isn't completely trivial and it would need to be redone anyway.

CAM-Gerlach commented on .github/ISSUE_TEMPLATE.md:
```diff
@@ -20,9 +20,9 @@
 
-1. 
-2. 
-3. 
```
As I mentioned above (specifically about this file), the way this is done is by adding the file to the blacklist in the script; its on a per-file basis. As such, I will add this file to the blacklist.

ccordoba12 replied:
Ok, that's even better.

CAM-Gerlach commented on .github/ISSUE_TEMPLATE.md:
```diff
@@ -42,11 +42,11 @@ PASTE TRACEBACK HERE
-* Python version: 
-* Qt version: 
-* PyQt version: 
-* Operating System name/version: 
```
See above.

CAM-Gerlach commented on scripts/rm_whitespace.py:
```diff
@@ -0,0 +1,291 @@
+#!/usr/bin/env python
```
Will do.

CAM-Gerlach commented on spyder/defaults/defaults-2.4.0.ini:
```diff
@@ -24,7 +24,7 @@ vertical_tabs = False
 use_custom_margin = True
```
> Revert all changes to this file 

See above. I'll add ``spyder/defaults/.*.ini``, to ignore all ``ini`` files in the defaults directly.

> It makes no sense to check for blanks in those files.

Why not?

That would miss other non-Spyder ``ini`` files, e.g. ``pytest.ini`` in the repo root. If someone injects spurious trailing spaces into that file, they would not get removed as they should by this script.

ccordoba12 replied:
> Why not?

I meant ini files for Spyder preferences.

CAM-Gerlach replied:
I see, I searched the repo and they are all in this dir, so the above regex should take care of them.

CAM-Gerlach commented on spyder/defaults/defaults-3.0.0.ini:
```diff
@@ -24,7 +24,7 @@ vertical_tabs = False
 use_custom_margin = True
```
See above.

CAM-Gerlach commented on spyder/plugins/ipythonconsole/widgets/shell.py:
```diff
@@ -209,8 +209,8 @@ def long_banner(self):
         if (pylab_o and sympy_o):
             lines = """
-Warning: pylab (numpy and matplotlib) and symbolic math (sympy) are both 
-enabled at the same time. Some pylab functions are going to be overrided by 
```
The script cannot currently ignore individual lines, and doing so would be extremely fragile. Furthermore, if we leave significant trailing spaces, our pre-commit hooks, Spyder's automatic trailing space removal, etc. will remove them if this file is ever touched, causing an undesired functional change (which is the very point of doing all this to begin with).

Therefore, I will instead modify these two lines to return the same output without trailing spaces.

CAM-Gerlach replied:
Actually, is there a reason this whitespace shouldn't be removed? It makes no difference to the output as they are just trailing space before line breaks (that serve no purpose), just like in the source file itself:

With:

![image](https://user-images.githubusercontent.com/17051931/58375937-3ff3c000-7f24-11e9-96c4-5023665f218c.png)

Without:

![image](https://user-images.githubusercontent.com/17051931/58375965-ed66d380-7f24-11e9-922f-a84cbce8e659.png)

ccordoba12 replied:
I thought it was necessary for the text to render well. If not, then it's fine to remove it.

CAM-Gerlach commented:
@ccordoba12 Thanks for your review; blacklisting those files and re-running now. Can ``create_app.py`` be moved to ``tools`` as well, or does it belong in ``scripts`` or the root (as it is now)?

CAM-Gerlach commented:
Temporarily closing this so the script doesn't pick it up.

pep8speaks commented:
Hello @CAM-Gerlach! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:





* In the file [`scripts/rm_whitespace.py`](https://github.com/spyder-ide/spyder/blob/f42c207e386822181cac96350cb2db32b78291d4/scripts/rm_whitespace.py):

> [Line 291:1](https://github.com/spyder-ide/spyder/blob/f42c207e386822181cac96350cb2db32b78291d4/scripts/rm_whitespace.py#L291): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file



















* In the file [`spyder/plugins/console/utils/interpreter.py`](https://github.com/spyder-ide/spyder/blob/f42c207e386822181cac96350cb2db32b78291d4/spyder/plugins/console/utils/interpreter.py):

> [Line 336:1](https://github.com/spyder-ide/spyder/blob/f42c207e386822181cac96350cb2db32b78291d4/spyder/plugins/console/utils/interpreter.py#L336): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file



* In the file [`spyder/plugins/console/widgets/shell.py`](https://github.com/spyder-ide/spyder/blob/f42c207e386822181cac96350cb2db32b78291d4/spyder/plugins/console/widgets/shell.py):

> [Line 819:5](https://github.com/spyder-ide/spyder/blob/f42c207e386822181cac96350cb2db32b78291d4/spyder/plugins/console/widgets/shell.py#L819): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 819:5](https://github.com/spyder-ide/spyder/blob/f42c207e386822181cac96350cb2db32b78291d4/spyder/plugins/console/widgets/shell.py#L819): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)





* In the file [`spyder/plugins/editor/widgets/tests/test_autocolon.py`](https://github.com/spyder-ide/spyder/blob/f42c207e386822181cac96350cb2db32b78291d4/spyder/plugins/editor/widgets/tests/test_autocolon.py):

> [Line 50:1](https://github.com/spyder-ide/spyder/blob/f42c207e386822181cac96350cb2db32b78291d4/spyder/plugins/editor/widgets/tests/test_autocolon.py#L50): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1







* In the file [`spyder/plugins/explorer/widgets.py`](https://github.com/spyder-ide/spyder/blob/f42c207e386822181cac96350cb2db32b78291d4/spyder/plugins/explorer/widgets.py):

> [Line 380:33](https://github.com/spyder-ide/spyder/blob/f42c207e386822181cac96350cb2db32b78291d4/spyder/plugins/explorer/widgets.py#L380): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent





* In the file [`spyder/plugins/io_hdf5/plugin.py`](https://github.com/spyder-ide/spyder/blob/f42c207e386822181cac96350cb2db32b78291d4/spyder/plugins/io_hdf5/plugin.py):

> [Line 16:80](https://github.com/spyder-ide/spyder/blob/f42c207e386822181cac96350cb2db32b78291d4/spyder/plugins/io_hdf5/plugin.py#L16): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)















* In the file [`spyder/plugins/variableexplorer/tests/test_variableexplorer.py`](https://github.com/spyder-ide/spyder/blob/f42c207e386822181cac96350cb2db32b78291d4/spyder/plugins/variableexplorer/tests/test_variableexplorer.py):

> [Line 30:80](https://github.com/spyder-ide/spyder/blob/f42c207e386822181cac96350cb2db32b78291d4/spyder/plugins/variableexplorer/tests/test_variableexplorer.py#L30): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)























* In the file [`spyder/utils/debug.py`](https://github.com/spyder-ide/spyder/blob/f42c207e386822181cac96350cb2db32b78291d4/spyder/utils/debug.py):

> [Line 147:1](https://github.com/spyder-ide/spyder/blob/f42c207e386822181cac96350cb2db32b78291d4/spyder/utils/debug.py#L147): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file

























* In the file [`spyder/utils/tests/test_sourcecode.py`](https://github.com/spyder-ide/spyder/blob/f42c207e386822181cac96350cb2db32b78291d4/spyder/utils/tests/test_sourcecode.py):

> [Line 40:31](https://github.com/spyder-ide/spyder/blob/f42c207e386822181cac96350cb2db32b78291d4/spyder/utils/tests/test_sourcecode.py#L40): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 40:34](https://github.com/spyder-ide/spyder/blob/f42c207e386822181cac96350cb2db32b78291d4/spyder/utils/tests/test_sourcecode.py#L40): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 40:46](https://github.com/spyder-ide/spyder/blob/f42c207e386822181cac96350cb2db32b78291d4/spyder/utils/tests/test_sourcecode.py#L40): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 40:53](https://github.com/spyder-ide/spyder/blob/f42c207e386822181cac96350cb2db32b78291d4/spyder/utils/tests/test_sourcecode.py#L40): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 42:33](https://github.com/spyder-ide/spyder/blob/f42c207e386822181cac96350cb2db32b78291d4/spyder/utils/tests/test_sourcecode.py#L42): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 42:36](https://github.com/spyder-ide/spyder/blob/f42c207e386822181cac96350cb2db32b78291d4/spyder/utils/tests/test_sourcecode.py#L42): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 42:48](https://github.com/spyder-ide/spyder/blob/f42c207e386822181cac96350cb2db32b78291d4/spyder/utils/tests/test_sourcecode.py#L42): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 42:55](https://github.com/spyder-ide/spyder/blob/f42c207e386822181cac96350cb2db32b78291d4/spyder/utils/tests/test_sourcecode.py#L42): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','















* In the file [`spyder/widgets/findreplace.py`](https://github.com/spyder-ide/spyder/blob/f42c207e386822181cac96350cb2db32b78291d4/spyder/widgets/findreplace.py):

> [Line 390:80](https://github.com/spyder-ide/spyder/blob/f42c207e386822181cac96350cb2db32b78291d4/spyder/widgets/findreplace.py#L390): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)

* In the file [`spyder/widgets/onecolumntree.py`](https://github.com/spyder-ide/spyder/blob/f42c207e386822181cac96350cb2db32b78291d4/spyder/widgets/onecolumntree.py):

> [Line 217:1](https://github.com/spyder-ide/spyder/blob/f42c207e386822181cac96350cb2db32b78291d4/spyder/widgets/onecolumntree.py#L217): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file

ccordoba12 commented:
> Can create_app.py be moved to tools as well

Yes, you can move it there.

> Also, any plans to merge #9226 soon?

I don't understand what's the rush. Can't your script be run again to get rid of the blanks in the files touched by that PR?

----

Also please take of the errors reported by pep8speaks that involve blanks.

ccordoba12 commented:
@CAM-Gerlach, what happened here?

CAM-Gerlach commented:
>  what happened here?

Same as when I closed it before,

> Temporarily closing this so the script doesn't pick it up.

Otherwise, the script would pick up all the files in this PR too, and it was faster and cleaner to close it then add a bespoke exception in the script itself. I didn't repeat that comment again this time since I'd already made it once when I did the same thing previously, which I (mistakenly) thought you would have seen.

> Yes, you can move it there.

Sure, done.

> I don't understand what's the rush.

There's no rush. I just figured it would be easier on both of us to get as many files as possible done in one PR, so we wouldn't have to deal with too many incremental followup PRs, but its fine.

> Also please take care of the errors reported by pep8speaks that involve blanks.

Do you mean all the errors with the word "blank" in them, all of them period, or something more specific? I implemented the first here but if you meant something else, let me know and I'll do that instead.

goanpeca commented:
probably

>Temporarily closing this so the script doesn't pick it up.

CAM-Gerlach commented:
@ccordoba12 Re-ran with files/paths added to blacklist, refactored blacklist to make adding items easier, added more robust and friendlier error handling at several points, automatically installed ``pr`` and ``pr-clean`` git aliases if they aren't present, improved documentation, removed blank line PEP 8 issues, general cleanup. Ready for final review.

ccordoba12 commented:
> I just figured it would be easier on both of us to get as many files as possible done in one PR

Ok, I understand. But I'm afraid this is going to require several PRs.

> Do you mean all the errors with the word "blank" in them, all of them period, or something more specific?

Yes, the errors with the word `blank` and the issues reported in `spyder/utils/tests/test_sourcecode.py`

CAM-Gerlach commented:
> But I'm afraid this is going to require several PRs.

Yeah, I knew it would require more than one, just wanted to minimize the number necessary. But its fine to merge it now.

> Yes, the errors with the word blank and the issues reported in spyder/utils/tests/test_sourcecode.py

The issues with the latter (missing a space after the commas) are no more germane to this PR than the other remaining ones (missing a space after ``#``, too much leading whitespace), so I just fixed all of them except for the line length (which would require more complex/mulitline changes).

ccordoba12 commented on CHANGELOG.md:
```diff
@@ -446,7 +446,7 @@ In this release 246 pull requests were closed.
 * [Issue 5763](https://github.com/spyder-ide/spyder/issues/5763) - Make arrow-key selection in multiple-option tab-completion dialogs "roll over" ([PR 5771](https://github.com/spyder-ide/spyder/pull/5771))
```
Please revert the changes to this file and put it in the excluded list. The blanks here come from blanks left in the title of PRs and issues, so I don't want to keep fixing them just for the sake of not avoiding blanks.

CAM-Gerlach replied:
I'm not sure I understand the point, when it takes zero extra effort to have the script remove them along with the others, they still shouldn't be there, they take up extra space in the file and our future commit hook will remove future spaces automatically (or we could just make sure loghub ``.strip()``s issue/PR titles).

Relative to that zero cost and zero downsides, it would require re-writing multiple commits to add it to the blacklist, test that it works, and revert adding the file, for no apparent gain. I'd rather spend that time adding ``.strip()`` to loghub and fixing the problem properly.

ccordoba12 replied:
I really think this is going too far. I agree with preventing to add blanks to code files, but this is not code. So, as you did with the Github template files, please add this one to the blacklist.

goanpeca replied:
Agreed this is going great so far, but we should really stick to code files only at the moment. Could you open an issue over loghub to fix the spacing issue? Thanks

CAM-Gerlach replied:
I still don't see a logically articulated reason for that distinction; either Markdown counts as code, or it counts as documentation, the latter of which we've removed spaces from for years, and regardless, they do not change the output. Its just extra work for no apparent gain to exclude it, but not enough to make arguing over it further worthwhile, so I made the change you requested.

For these purposes, there's little point fixing the issue with loghub if we don't fix it here, since there will still be spurious whitespace left in the file if we don't clean them up. Why should we care about loghub not stripping trailing spaces if we don't actually care about them in the output?
